src/peft/tuners/tuners_utils.py

```diff
@@ -486,30 +484,43 @@ def check_target_module_exists(config, key: str) -> bool | re.Match[str] | None:
         None if no match found
     """
     if isinstance(config.target_modules, str):
-        target_module_found = re.fullmatch(config.target_modules, key)
+        target_module_found = (config.target_modules == key) or re.fullmatch(config.target_modules, key)
+    elif key in config.target_modules:
+        # this module is specified directly in target_modules
+        target_module_found = True
     else:
-        target_module_found = key in config.target_modules or any(
-            key.endswith(f".{target_key}") for target_key in config.target_modules
-        )
-        is_using_layer_indexes = getattr(config, "layers_to_transform", None) is not None
-        layer_indexing_pattern = getattr(config, "layers_pattern", None)
+        target_module_found = any(key.endswith(f".{target_key}") for target_key in config.target_modules)
 
+        layer_indexes = getattr(config, "layers_to_transform", None)
+        layers_pattern = getattr(config, "layers_pattern", None)
+
+        is_using_layer_indexes = layer_indexes is not None and (
+            len(layer_indexes) != 0 if isinstance(layer_indexes, list) else True
+        )
         if is_using_layer_indexes and target_module_found:
-            layers_pattern = COMMON_LAYERS_PATTERN if layer_indexing_pattern is None else layer_indexing_pattern
-            layers_pattern = [layers_pattern] if isinstance(layers_pattern, str) else layers_pattern
-
-            for pattern in layers_pattern:
-                layer_index = re.match(f".*.{pattern}\.(\d+)\.*", key)
-                if layer_index is not None:
-                    layer_index = int(layer_index.group(1))
-                    if isinstance(config.layers_to_transform, int):
-                        target_module_found = layer_index == config.layers_to_transform
-                    else:
-                        target_module_found = layer_index in config.layers_to_transform
-
-                    break
+            # to keep up with the previous results(if layers_pattern is an empty list, ret is True)
+            # and make the result consistent for all empty layers_pattern: None, [], ""
+            # COMMON_LAYERS_PATTERN is dropped
+            # empty layers_pattern means all layer pattern is ok
+            layer_index = None
+            if layers_pattern is None or len(layers_pattern) == 0:
+                layer_index = re.match(r".*\.[^.]*\.(\d+)\.", key)
```

Comment:
Why would you say it's better not to rely on `COMMON_LAYERS_PATTERN` here? If we decide it's really not needed, it should be completely removed from the PEFT code base.

Reply:
I didn't find any other direct usage of COMMON_LAYERS_PATTERN, so I think it won't be harmful. As for the reason why I drop it, I try to avoid changing the output and keep the logic consistent. In the previous code, if the layers_pattern is an empty list, all layer indexes can be matched, as stated in https://github.com/huggingface/peft/blob/482a2a6d9aaa01d534b1240e8c1ab6d346eb278f/tests/test_tuners_utils.py#L59C9-L59C9
In fact, COMMON_LAYERS_PATTERN also doesn't work here. Maybe this is a bug, but I don't change it to keep the output consistent when layers_pattern is an empty list. And now '[]' makes all layers being matched, I think None(default value) should follow this too, even more suitable than '[]'. So I just make them all consistent, None, [], ""(empty string). Then empty layers_pattern means any layer pattern is ok, it's also reasonable. Maybe there should be a warning after the layers_pattern is set.
And torch.nn.ModuleList where use numeric value to be the module name when add modules, i.e. the layer index, so I don't think others will use numeric value as their own module name. Also there can't be '.' in module name, so I think it's safe to just match "[^.]*\.\d+\.".
How to treat empty list in layers_pattern is controversial, here I just follow previous 'bug'.